### PR TITLE
Add support for tracking ERC20 token transfers

### PR DESCRIPTION
**Pull request by maxwellfoley3**
Working on adding events for ERC20 token transfers, not yet finished

### CODE REVIEW

File: scripts/listener.ts [hunk 1]
```diff
@@ -95,6 +98,21 @@ class StandaloneEventHandler extends IEventHandler {
 
 const skipCatchup = false;
 
+const tokenListUrls = [
+  "https://wispy-bird-88a7.uniswap.workers.dev/?url=http://tokenlist.aave.eth.link",
+  "https://gateway.ipfs.io/ipns/tokens.uniswap.org",
+  "https://wispy-bird-88a7.uniswap.workers.dev/?url=http://defi.cmc.eth.link"
+]
+
+async function getTokenLists() {
```
Comment (jnaviask) [inline scripts/listener.ts:107]: Please lint this function/the rest of this file. We should add it to the lint files list tbh.

File: src/erc20/filters/labeler.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  data: IEventData
+): IEventLabel => {
+  switch (data.kind) {
+    case EventKind.Approval:
+      return {
+        heading: 'Approval',
+        label: `Owner ${fmtAddr(data.owner)} approved spender ${fmtAddr(
+          data.spender
+        )}
+        to spend ${data.value}`,
```
Comment (jnaviask) [inline src/erc20/filters/labeler.ts:26]: Add a period here for consistency.

File: src/erc20/filters/labeler.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      };
+    case EventKind.Transfer:
+      return {
+        heading: 'Transfer',
+        label: `Transfer of ${data.value} from ${data.from} to ${data.to}.`,
```
Comment (jnaviask) [inline src/erc20/filters/labeler.ts:31]: We may also want to figure out how to format these values in terms of the token symbol, if possible. Likely via optional arg that we pass through.

File: src/erc20/filters/titler.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * This a titler function, not to be confused with the labeler -- it takes a particular
+ * kind of event, and returns a "plain english" description of that type. This is used
+ * on the client to present a list of subscriptions that a user might want to subscribe to.
+ */
+export const Title: TitlerFilter = (kind: EventKind): IEventTitle => {
+  switch (kind) {
+    case EventKind.Approval:
+      return {
+        title: 'Delegation Approved',
+        description: 'One account delegated tokens to another..',
```
Comment (jnaviask) [inline src/erc20/filters/titler.ts:14]: We should update the copy here -- this gets displayed at the "subscriptions" page -- but currently that page isn't active, so it's not a big deal.

File: src/erc20/subscribeFunc.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+export const subscribeEvents: SubscribeFunc<
+  Api,
+  RawEvent,
+  IErc20SubscribeOptions
+> = async (options) => {
+  const { chain, api, handlers, verbose } = options;
+  // helper function that sends an event through event handlers
+  const handleEventFn = async (event: CWEvent<IEventData>): Promise<void> => {
+    let prevResult = null;
+  for (const handler of handlers) {
```
Comment (jnaviask) [inline src/erc20/subscribeFunc.ts:100]: Looks like some tabbing inconsistencies here?

File: src/erc20/subscriber.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  public async addNewToken(
+    tokenAddress: string,
+    retryTimeMs = 10 * 1000,
+    retries = 5
+  ): Promise<void> {
+    const existingToken = this.api.tokens.find((o) => {
+      return o.address === tokenAddress;
+    });
+    if (existingToken) {
+      return; // Token is already being monitored
```
Comment (jnaviask) [inline src/erc20/subscriber.ts:60]: I would add a log.info statement here noting that we're ignoring the new token.

File: src/erc20/types.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  from: Address;
+  to: Address;
+  value: number;
+  contractAddress: Address;
+}
+
+export type IEventData = IApproval | ITransfer;
+// eslint-disable-next-line semi-style
+
+export class Token {
```
Comment (jnaviask) [inline src/erc20/types.ts:51]: Where does this get used?

**Review by jnaviask — CHANGES_REQUESTED** [review]
Code looks fine besides a few small formatting and copy changes. Will QA it next.

**Comment by jnaviask** [comment]
Might want to add a few more flags to the listener:
* to allow listening to a single contract (via the -c flag)
* to support the dotenv file
* to have a default (infura) url for erc20 like the others

But not a big deal. Core functionality seems fine, can merge when the above changes go in.

File: src/erc20/filters/titler.ts (hunk 4 above)
Comment (maxwellfoley3) [reply src/erc20/filters/titler.ts:14]: Update it to what?

File: src/erc20/types.ts (hunk 7 above)
Comment (maxwellfoley3) [reply src/erc20/types.ts:51]: The `Token` class gets used in the `Subscriber` class, if that's what you're asking

File: src/erc20/filters/titler.ts (hunk 4 above)
Comment (jnaviask) [reply src/erc20/filters/titler.ts]: Maybe just remove the double period... we can deal with updating the actual copy later.

**Pull request merged**